### ISSUE TITLE
VSCode extension

### ISSUE THREAD

**Issue by baubleb** [sile-typesetter/sile#1406]
Currently many use integrated editor environments for LaTeX, for example TexStudio.

It would be nice to have an extenson for VSCode for syntax highlighting (and other features over time such as document *preview*, *intellisense* and *snippets*) for SILE to make it a better writing experience.

Further information on VSCode extensions, here

[Extension API](https://code.visualstudio.com/api)
[Extension Anatomy](https://code.visualstudio.com/api/get-started/extension-anatomy)
[Syntax Highlighting Guide](https://code.visualstudio.com/api/language-extensions/syntax-highlight-guide)
[Code Completion](https://code.visualstudio.com/api/language-extensions/programmatic-language-features#show-code-completion-proposals)

**Comment by alerque** [comment]
I agree there is a need and would be happy to facilitate this. If somebody wants to work on it and gets at least a proof of concept going that does *something* I'd be happy to host it in the GitHub org here and get whoever is working on it the necessary permissions. That being said I don't use VSCode myself so I likely won't be cooking up an extension on my own—somebody else with the expertise and the itch will have to tackle this.

For my own part I use NeoVIM and hence we already have a [`vim-sile`](https://github.com/sile-typesetter/vim-sile) plugin with full syntax highlighting support.

Not 100% the same issue but as a related side note, the one thing I would like to see a VSCode extension for is for importing the syntax highlighting rules into [linguist](https://github.com/github/linguist/) and others so we can have syntax highlighting here on GitHub for sil sources. For that we also need a threshold number of sil files posted in repos "in the wild". Last I looked we were about half way there. If you have open licensed projects and/or even sandbox testing and experiments that *can* be posted publicly...

**Comment by Omikhleia** [comment]
FWIW, I use LaTeX syntax coloring and the [LTeX](https://marketplace.visualstudio.com/items?itemName=valentjn.vscode-ltex) VSCode extension as basic grammar/spell checker. It works "decently" with SIL files, although not perfectly...

(As for document preview, I've no idea but it might be a bit hard when working with included files, no?)

**Comment by baubleb** [comment]
A SILE Language Server might help here, it then could be used by many writing environments (IDE's) that are language server compatible.

[Language Server Protocol](https://microsoft.github.io/language-server-protocol/)

This would be more widely benefitial for more writing environments than a port/modification of TexStudio to be SileStudio.

VSCode is a jack of all trades IDE and widely used.

**Comment by alerque** [comment]
Yes, an LSP language server for the `sile` format would be great. In that case some integration with Lua LSPs to expose the SILE API for completion and in-editor documentation would be amazing.

Again I don't expect to be able to work on this soon myself, but would gladly chip in to facilitate it if somebody wants to get started.

**Comment by colbyn** [comment]
I've been working on my own markup language and the syntax is pretty similar to SILE. For instance it seems like dashes and colons as valid cmd identifiers in SILE. 

Unless I'm mistaken?

Anyway if it's useful, here's an example
<img width="597" alt="Screen Shot 2022-09-30 at 10 35 18 PM" src="https://user-images.githubusercontent.com/2118601/193392766-8f8bb9e2-b82e-4ee6-abdd-68941a3412c3.png">

Which breaks with LaTeX syntax highlighting
<img width="644" alt="Screen Shot 2022-09-30 at 10 35 30 PM" src="https://user-images.githubusercontent.com/2118601/193392774-966ba75a-f41a-4bb7-8ca4-20789c96a550.png">

Anyway it's a pretty basic implementation at the moment, but if it's useful, you can find the **vscode** implementation [over here](https://github.com/subscript-publishing/subscript/tree/master/ss-language-server/vscode-subscript-markup-language)...

**Comment by alerque** [comment]
[This branch](https://github.com/alerque/sile/compare/formal-grammar...alerque:sile:ebnf) has an EBNF grammar, which might be useful for generating a treesitter parser that would work in VSCode and others for syntax highlighting. Input on how to get tooling spun up derived from this or for fixing mistakes in the grammar would be welcome.

**Comment by Omikhleia** [comment]
Greetings,

This issue hasn't seen any contribution proposal for several years, and hasn't been discussed for more than a year...

The branch mentioned by @alerque, with ABNF and EBNF grammars of the SIL language, was merged to the [website](https://sile-typesetter.org/sil-grammar/) some time ago, with nice pictures and good explanations. Potential contributors have thus been given the needed information there to possibly proceed and spin something off...

As I stated above, I too would have been interested in a VSCode-compatible LSP for the SIL TeX-like syntax at one point. It didn't happen, and we should also note the following facts on that topic:
 - Here, we are in the SILE typesetter repository, i.e. focused on the core capabilities of the SILE engine. As a community, discussing such "external" add-ons for editors is of course welcome for a start, but eventually such add-ons would be projects of their own, with their own repository, lifecycle, issue tracking and release plans... So having a long-standing issue on a topic that doesn't directly relate to the core typesetting engine is annoying...
 - SIL TeX-like is but one of the input markup languages available to users. There are good LSPs for XML, Markdown -- and perhaps eventually for Djot; so the appeal for original issue (which didn't mention the SIL TeX-like syntax flavor explicitly, but that was likely inferred...) is somewhat of a niche, which may explain why the contributors to the discussion didn't go farther.

Therefore, I am adding the "pending closure" label to this issue, as part of a backlog cleaning pass and triage attempt.
Interested users are invited to say their word, and propose their strategies and involvement for the issue to be addressed. If nothing occurs, it will be rejected after some delay, without further notice.
